A110-child-channel-plugins.md

```diff
@@ -0,0 +1,215 @@
+# **A110: Child Channel Configuration Plugins**
+
+* **Author(s)**: [Abhishek Agrawal](mailto:agrawalabhi@google.com)
+* **Status**: Draft
+* **To be implemented in**: Core, Java, Go
+* **Last updated**: 2025-12-23
+
+## **Abstract**
+
+This proposal introduces a mechanism to configure "child channels"—channels created internally by gRPC components (such as xDS control plane channel). Currently, these internal channels cannot easily inherit configuration (like metric sinks and interceptors) from the user application without relying on global state. This design proposes a language-specific approach to injecting configuration: using `functional interfaces` in Java, `DialOptions` in Go, and `ChannelArgs` in Core.
+
+## **Background**
+
+Complex gRPC ecosystems often require the creation of auxiliary channels that are not directly instantiated by the user application. Two primary examples are:
+
+1. **xDS (Extensible Discovery Service)**: When a user creates a channel with an xDS target, the gRPC library internally creates a separate channel to communicate with the xDS control plane. Currently, users have limited ability to configure this internal control plane channel.
+2. **Advanced Load Balancing (RLS, GrpcLB):** Policies like RLS (Route Lookup Service) and GrpcLB, as well as other high-level libraries built on top of gRPC, frequently create internal channels to communicate with look-aside load balancers or backends.
+
+### **The Problem**
+
+There is currently no standardized way to configure behavior for these child channels.
+
+* **Metrics**: Users need to configure metric sinks so that telemetry from internal channels can be read and exported.
+* **Interceptors**: Users may need to apply specific interceptors (e.g., for authentication, logging, or tracing) to internal traffic.
+* **No Global State**: These configurations cannot be set globally (e.g., using static singletons) because different parts of an application may require different configurations, such as different metric backends or security credentials.
+
+## **Proposal**
+
+The proposal creates a "plugin" or configuration injection style for internal channels. The implementation varies by language to match existing idioms, but the goal remains consistent: allow the user to pass a configuration object or function that the internal channel factory applies during creation.
+
+### **Java**
+
+In Java, the configuration will be achieved by accepting functions (callbacks). The API allows users to pass a `Consumer<ManagedChannelBuilder<?>>` (or a similar functional interface). When an internal library (e.g., xDS, RLS, gRPCLB) creates a child channel, it applies this user-provided function to the builder before building the channel.
```

Comment:
The API proposed for Java and Go seems to be only for channels, not for servers.  However, the xDS, ext_authz, and ext_proc cases will also happen on the gRPC server side.  Do we need to solve this problem also on the server side?

(The C-core approach will work on both client and server sides.)